src/core/hle/kernel/thread.cpp

```diff
@@ -3,433 +3,972 @@
+        return std::find_if(waiting_on.begin(), waiting_on.end(), [&](const auto& obj) {
+                   return obj.get() == object;
+               }) != waiting_on.end();
+
+    case Status::WaitSyncAll:
+        return std::find_if(waiting_on.begin(), waiting_on.end(), [&](const auto& obj) {
+                   return obj.get() == object || !obj->ShouldWait(this);
+               }) == waiting_on.end();
+
+    default:
+        return false;
+    }
+}
+
+void Thread::ResumeFromWait(Kernel::WaitObject* object) {
+    switch (status) {
+
+    case Status::WaitSyncAll:
+        for (auto obj : waiting_on) {
+            obj->Acquire(this);
         }
-        owner_process->memory_used += Memory::PAGE_SIZE;
+        SyncOutput(RESULT_SUCCESS);
+        break;
 
-        tls_slots.emplace_back(0); // The page is completely available at the start
-        available_page = tls_slots.size() - 1;
-        available_slot = 0; // Use the first slot in the new page
+    case Status::WaitSyncAny:
+    case Status::WaitIPC: {
+        auto it = std::find_if(waiting_on.begin(), waiting_on.end(),
+                               [&](const auto& obj) { return obj.get() == object; });
+        const auto idx = static_cast<s32>(it - waiting_on.begin());
+        object->Acquire(this);
+
+        if (status == Status::WaitIPC && object->GetHandleType() == HandleType::ServerSession) {
+            auto server_session = static_cast<ServerSession*>(object);
+            auto result = server_session->ReceiveIPCRequest(this);
+            SyncOutput(result, idx);
+        } else {
+            SyncOutput(RESULT_SUCCESS, idx);
+        }
+        break;
+    }
 
-        auto& vm_manager = owner_process->vm_manager;
+    case Status::WaitHleEvent:
+        object->Acquire(this);
+        hle_context->OnWakeUp(this, WakeupReason::Signal, hle_callback);
+        hle_context = nullptr;
+        hle_callback = nullptr;
 
-        // Map the page to the current process' address space.
-        vm_manager.MapBackingMemory(Memory::TLS_AREA_VADDR + available_page * Memory::PAGE_SIZE,
-                                    memory.GetFCRAMRef(*offset), Memory::PAGE_SIZE,
-                                    MemoryState::Locked);
+    default:
+        ASSERT_MSG(false, "Inconsistent thread status");
     }
 
-    // Mark the slot as used
-    tls_slots[available_page].set(available_slot);
-    thread->tls_address = Memory::TLS_AREA_VADDR + available_page * Memory::PAGE_SIZE +
-                          available_slot * Memory::TLS_ENTRY_SIZE;
+    // Clear waiting objects/threads
+    for (auto obj : waiting_on) {
+        obj->RemoveWaitingThread(this);
+    }
+    waiting_on.clear();
+    core.root.UnscheduleEvent(wakeup_event.get());
+    status = Thread::Status::Ready;
+}
 
-    memory.ZeroBlock(*owner_process, thread->tls_address, Memory::TLS_ENTRY_SIZE);
+void Thread::SetPriority(u32 priority) {
+    nominal_priority = priority;
+    for (auto obj : waiting_on) {
```

Comment:
```suggestion
    for (auto& obj : waiting_on) {
```

otherwise this can cause atomic reference count increment/decrement churn.